### ISSUE TITLE
Synchronized colors in the console

### ISSUE THREAD

**Issue by svick** [dotnet/corefx#2808]
TLDR: I believe that there is currently no way to correctly write colored messages to the console in a multi-threaded environment. There should be one.

Long version:

When you want to write a colored message to the console, the common pattern looks something like:

``` c#
Console.ForegroundColor = ConsoleColor.Red;
Console.WriteLine("ERROR!");
Console.ResetColor();
```

This approach is not thread-safe, since other writers can also write while the color is set.

An improvement is [to lock on a private object](http://stackoverflow.com/a/13861466/41071):

``` c#
static object lockObj = new object();

...

lock (lockObj)
{
    Console.ForegroundColor = ConsoleColor.Red;
    Console.WriteLine("ERROR!");
    Console.ResetColor();
}
```

This will synchronize correctly among writers that use the same lock object, but other writers can still end up with incorrect color.

On the Desktop .Net, writing to the console uses [`SyncTextWriter`](http://referencesource.microsoft.com/#mscorlib/system/io/textwriter.cs,9e3dd0323cf4ee8b) as `Console.Out` which locks on itself. This means that locking on `Console.Out` will synchronize correctly with other non-colored writers (those that just call `Console.WriteLine` without any additional synchronization) and colored writers which also lock on `Console.Out`:

``` c#
lock (Console.Out)
{
    Console.ForegroundColor = ConsoleColor.Red;
    Console.WriteLine("ERROR!");
    Console.ResetColor();
}
```

On CoreFX, [a different version of `SyncTextWriter`](https://github.com/dotnet/corefx/blob/fa6e6a1e6300cdbffa31998a8c70d3bb9b7d06c3/src/System.Console/src/System/IO/SyncTextWriter.cs) is used, which locks on its own private lock object. This is generally considered a better practice than locking on `this`, but it also means there is no way to write a colored message to the console that is correctly synchronized with non-colored writers.

It would be nice if there was an officially supported pattern to write a colored message to the console with correct synchronization on both Desktop and Core. Failing that, I think that `SyncTextWriter` on Core should be implemented the same way as the Desktop version, so that the `lock (Console.Out)` version also worked on Core, event if it's relying on an implementation detail.

**Comment by svick** [comment]
CC: @stephentoub, I believe you worked on Console in the past.

**Comment by stephentoub** [comment]
Interesting.  Thanks for raising this, @svick.

@pallavit, thoughts on this?  Looking back through history in TFS, I believe this changed as part of overhauling the standalone System.Console library for .NET Core (Changeset 1303167).

I'm a fairly staunch advocate for not locking on exposed instances.  However, @svick makes a good case, and based on this I think switching back to locking on ```this``` is the lesser of two evils.  There's a pattern that can be used on desktop that fails to work correctly with .NET Core, and there's no substitute way of achieving that support on .NET Core.

**Comment by pallavit** [comment]
@svick Thanks for reporting the issue. @stephentoub Yes, I think we need to address this. Which other pattern were you considering which works on Desktop but not on .NetCore?

**Comment by stephentoub** [comment]
> Which other pattern were you considering which works on Desktop but not on .NetCore?

Sorry what I wrote wasn't clear... what I meant was that *this* pattern of locking on Console.Out works on desktop and fails to work on .NET Core, and there's no substitute for that on .NET Core that I'm aware of.

**Comment by pallavit** [comment]
Assigning the issue to myself :)

**Comment by JeremyKuhne** [comment]
This appears to still be broken. We don't lock for every method on TextWriter (a number of them are missing, see the various format WriteLine overloads).

Since this was originally addressed we've added TextWriter.Synchronized() which is what the original code used. https://referencesource.microsoft.com/#mscorlib/system/console.cs,364

I'm going to switch back to using TextWriter.Synchronized() and create a follow-up issue- we did not create synchronized overloads for the new Span APIs in TextWriter either.

**Comment by stephentoub** [comment]
@JeremyKuhne, this can be closed now, right?
